### PR TITLE
mgr: perf counters node exporter

### PR DESCRIPTION
**Pull request by avanthakkar**
The number of perf counters keeps growing in each ceph version and mgr's performance is fragile when moving big payloads of data around because of subinterpreters. A way to reduce this load is by moving perf counters away from the mgr. 

**update**: This PR will be the first of many which will be merge onto the feature branch `feature-exporter`. This PR is intended to have all features marked as done in _Next steps_ except TLS support.

**Next steps**:
- [x] List socket files and make use of them
- [x] Create http server to expose GET /metrics for prometheus
- [x] Load perf counters from all daemons and transform to prometheus format
- [x] Containarize
- [x] Introduce a exporter container for cephadm
- [ ] MetricCollector discovery in cephadm in order to configure prometheus to scrape all MetricCollector endpoints (https://github.com/ceph/ceph/pull/45425)
- [ ] Ensure rook support
- [x] Posiblity to expose perf counters so that modules can request through http/whatever
- [x] **Security**: HTTPS / TLS 🆕 
- [ ] **Security**: HTTP Basic Auth 🆕 
- [ ] **Ops**: configuration (via ceph.conf + mon config opts) 🆕 
- [ ] **Ops**: application log (errors, warnings, etc.) 🆕 
- [ ] **Ops**: server log (HTTP requests) 🆕 
- [ ] **Performance**: HTTP compression 🆕
- [ ] **Performance**: memory/cpu footprint, soak testing (mem-leak) 🆕
- [ ] **Documentation** 🆕
- [ ] **Tests**: Integration tests with teuthology
- [ ] **Tests**: Add unit testing

Proposal for node-exporter:
Screenshot from 2022-03-02 21-30-43

Signed-off-by: Pere Diaz Bou <pdiazbou@redhat.com>
Co-authored-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/msg/msg_types.h
Comment (cbodley) [inline src/msg/msg_types.h:54]: what's the motivation for a new entity type? can we just use `TYPE_CLIENT` here?

File: src/ceph_exporter.cc
Comment (cbodley) [inline src/ceph_exporter.cc:86]: why disable http keepalive here?

as a server, you should generally leave this up to the clients - they'll know whether or not they intend to send more requests on this connection. the clients communicate this in the [Connection](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection) header

you can check whether the client requested keep-alive using `request_.keep_alive()`. if that's true, then you should call back into `read_request()` after `write_response()`'s successful completion to handle the next request on this connection

File: src/mgr/DaemonMetricCollector.cc
Comment (cbodley) [inline src/mgr/DaemonMetricCollector.cc:14]: ok, this is the first time i've seen boost.json used in ceph. i tend to think we should stick with rapidjson, but boost.json is probably a lot easier to use for a prototype 👍

**Review by cbodley — COMMENTED** [review]
the beast stuff in `http_connection` looks good, nice work so far

**Comment by mattbenjamin** [comment]
I'm not clear whether the domain socket data interchange for counters with labels is well defined/finalized, or not.  If it is, I'd like to see if Vijay Deloiya can get started on work to make RGW export examples of that format.

**Comment by avanthakkar** [comment]
> @yuvalif At first we were thinking of trying golang because of ODF/OCP and prometheus support. Since making a http endpoint for /metrics and the format of metrics was simple enough we decided to craft it ourselves, and given the support for ceph daemon socket files (`AdminSocketClient`) existed, we went for it simply because of convenience.

Yes and to add also no extra dependencies required , boost/beast library already used in ceph.  Also this exporter maybe used for collecting metrics from ceph daemons in future so the messaging b/w them will be easier as they talk the same language.

**Comment by pereman2** [comment]
> > > @yuvalif At first we were thinking of trying golang because of ODF/OCP and prometheus support. Since making a http endpoint for /metrics and the format of metrics was simple enough we decided to craft it ourselves, and given the support for ceph daemon socket files (`AdminSocketClient`) existed, we went for it simply because of convenience.
> > 
> > 
> > Yes and to add also no extra dependencies required , boost/beast library already used in ceph. Also this exporter maybe used for collecting metrics from ceph daemons in future so the messaging b/w them will be easier as they talk the same language.
> 
> @pereman2 and @avanthakkar thanks for the reply! I also thought that golang might be more appropriate here (as this is a "control plane daemon"). I guess that this would require to re-implement the admin socket client in golang. which would be extra effort - do you have an estimate how much work would that be? also, so we have a stable API there, or is this often changing? BTW, I don't think that messaging bandwidth would change as long as we keep the same wire protocol between the ceph daemons and the exporter.

AdminSocketClient isn't complex to implement, but since Avan and me are unexperienced with golang I wouldn't know the estimate time to complete (I wouldn't mind learning golang tho). 

I'm not sure about how stable AdminSocketClient is but it doesn't seem like something that might change soon.

**Review by epuertat — COMMENTED** [review]
Really nice work @pereman2 @avanthakkar !! I haven't checked the http_server.cc file yet, but in order to give you guys feedback asap, there it goes!

File: src/CMakeLists.txt
Comment (pereman2) [reply src/CMakeLists.txt:923]: I've been testing this with vstart the whole time :P. Incremeting osds with `vstart --inc-osd`was successful also

File: src/common/entity_name.cc
Comment (pereman2) [reply src/common/entity_name.cc:23]: to_array is included in version 20. We are using 17 if I'm correct.

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm:48]: > Why a new container image? Isn't this package/binary included in the standard `ceph-ci/ceph` image?
> 
> OTOH, I like the idea of creating lightweight containers, but if that's the case, the final location for this should be in quay.io (and I'm not so sure that the RPM deps don't end up installing everything either).

Yeah it's bascially a quay image but it also installs the exporter package so I had to write a Dockerfile for it and thats why I'm using this new image. And FIY this is just for testing! Once we hv all required C++ changes merged in this feature branch we can have a new image which will be based on master , bcz then this new expoter package will be includes by default in all the quay images.

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm:848]: Yeah we need a similar spec for the exporter service so uit'll be pretty much same as other services

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm]: Already done! 9926 is ours!!

File: ceph.spec.in
Comment (avanthakkar) [reply ceph.spec.in:711]: Yeah I'll change the description, thanks!

File: src/exporter/DaemonMetricCollector.cc
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc]: @epuertat I did some refactoring on that. Since this function is more complex I'm not sure I like extracting to a lot of functions because they aren't being used more than one time. Anyways, I did it.

Comment (epuertat) [reply src/exporter/DaemonMetricCollector.cc]: Reuse is only one of the reasons to extract code to an external method: improving code quality (readability, cyclomatic complexity) or testability are good reasons too.

That code comprises a triple-nested for loop, so breaking that down to separate methods would allow you to improve unit test isolation (testing the inner methods without the overhead of outer iterations; testing the outer methods mocking the inner ones).

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm:5732]: Yeah it'll only install the daemon if skip_monitoring_stack is not True..just like other monitoring containers!
https://github.com/ceph/ceph/pull/45220/files#diff-d0f7acffbce59b9e36a1479d1b1f32955cd811ab8b0b1aa0018a2003f0c4d73eR5071

File: src/cephadm/cephadm
Comment (epuertat) [reply src/cephadm/cephadm:914]: Yup but the type hint is correct? Is this going to return a `CephIscsi` object or a `CephExporter`? If mypy didn't catch this, maybe there's an error with its config.

File: src/exporter/DaemonMetricCollector.cc
Comment (epuertat) [inline src/exporter/DaemonMetricCollector.cc:53]: Tabs look weird here tbh :/

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm]: Ah yes that was the lint issue..its fixed now!

File: ceph.spec.in
Comment (avanthakkar) [reply ceph.spec.in]: Done! How about now? ("Daemon for exposing perf counters as Prometheus metrics")

Comment (epuertat) [reply ceph.spec.in]: Much better! Thank you @avanthakkar !

File: src/msg/msg_types.h
Comment (avanthakkar) [reply src/msg/msg_types.h]: It's no more needed

File: src/exporter/DaemonMetricCollector.cc
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc]: Yeah I understand that. I'll add a new task to refactor this on the exporter doc so that we keep track of it. But for now, let's try to push this PR onto the feature branch.

File: src/exporter/DaemonMetricCollector.cc
Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc]: Formatted the files using [C/C++ ](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools) for VS Code

**Comment by avanthakkar** [comment]
> @adk3798 @rkachach Can you guys review the cephadm changes? Also I see a flake8 error , but not sure how to solve it..can you help?
> 
> ```
> flake8 installdeps: flake8
> flake8 installed: flake8==4.0.1,mccabe==0.6.1,pycodestyle==2.8.0,pyflakes==2.4.0
> flake8 run-test-pre: PYTHONHASHSEED='7413244'
> flake8 run-test: commands[0] | flake8 --config=tox.ini alerts balancer cephadm cli_api crash devicehealth diskprediction_local hello iostat localpool nfs orchestrator prometheus selftest
> flake8 run-test: commands[1] | bash -c 'test $(git ls-files cephadm | grep ".py$" | grep -v tests | xargs grep "docker.io" | wc -l) == 15'
> ERROR: InvocationError for command /usr/bin/bash -c 'test $(git ls-files cephadm | grep ".py$" | grep -v tests | xargs grep "docker.io" | wc -l) == 15' (exited with code 1)
> ```

I've fixed the flake8 error. It's just a temporary change, as in future once the feature branch is merged in master the exporter
would be included in default master ceph image

File: src/exporter/DaemonMetricCollector.cc
Comment (cbodley) [inline src/exporter/DaemonMetricCollector.cc:113]: we need to handle errors from `do_request()` here. we should expect admin sockets to go away randomly as daemons are started/stopped

on errors, it looks like `do_request()` will return a non-empty std::string with the error message

File: src/exporter/http_server.cc
Comment (cbodley) [inline src/exporter/http_server.cc:94]:
```c++
int main(int argc, char** argv) {
  boost::thread server_thread(http_server_thread_entrypoint);
  DaemonMetricCollector &collector = collector_instance();
  collector.main();
  server_thread.join();
}
```
with `http_server` and `DaemonMetricCollector` running in separate threads, don't we need to protect this access under a mutex?

File: ceph.spec.in
Comment (cbodley) [inline ceph.spec.in:711]: apologies for bike-shedding here, but i think a more specific name like `ceph-prometheus-exporter` would be preferable here

**Review by cbodley — COMMENTED** [review]
great progress here 👍

File: src/exporter/http_server.cc
Comment (avanthakkar) [reply src/exporter/http_server.cc:153]: Yes that's the plan for the next iteration, to introduce some configs like port, addrs, cert, key,etc.

File: src/exporter/http_server.cc
Comment (avanthakkar) [reply src/exporter/http_server.cc:155]: Yes, port will be be made configurable in the next iteration

File: src/exporter/DaemonMetricCollector.cc
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc:200]: much better

File: src/exporter/DaemonMetricCollector.cc
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc:179]: I think I'll keep a map since I want to store the daemon_name. Do you think that a `vector<pair<string, socket>>` might be better?

File: src/exporter/http_server.cc
Comment (pereman2) [reply src/exporter/http_server.cc:94]: Which access? The only thing I see that could need a mutex is DaemonMetricCollector::get_metrics

File: src/exporter/DaemonMetricCollector.cc
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc:17]: On a host, cephadm should deploy an exporter per cluster since containers mount their respective cluster asok files. 

As far as I know, cephadm has `/var/run/ceph/{fsid}` folders for eachs clusters. If you create a container on a cluster it will mount `/var/run/ceph/{fsid}` to`/var/run/ceph/`, therefore, the exporter if deployed for each cluster, should parse sockets related to the cluster he is in.

Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc:17]: Anyways, as we plan on creating a diff PR onto the feature branch with configuration files, we can add an option to change the asok folder location.

File: src/exporter/CMakeLists.txt
Comment (pereman2) [reply src/exporter/CMakeLists.txt:4]: Isn't it there?

File: src/exporter/DaemonMetricCollector.cc
Comment (cbodley) [reply src/exporter/DaemonMetricCollector.cc]: > Anyways, as we plan on creating a diff PR onto the feature branch with configuration files, we can add an option to change the asok folder location.

sounds good, thanks. we should try to support the non-containerized mode too, by filtering filenames in "$run_dir/$cluster-$name.asok" form by cluster prefix

File: src/exporter/DaemonMetricCollector.cc
Comment (cbodley) [reply src/exporter/DaemonMetricCollector.cc]: yeah map is fine, the point is just that we need a way to clean up the `AdminSocketClient` entries after their file goes away. calling `do_request()` on a deleted file should return an error, so we should handle errors by erasing that map entry

File: src/exporter/http_server.cc
Comment (cbodley) [reply src/exporter/http_server.cc:93]: right. doesn't that need a mutex?

File: src/exporter/DaemonMetricCollector.cc
Comment (pereman2) [reply src/exporter/DaemonMetricCollector.cc]: Then I mark this as resolved. It was fixed last in the latest commit

File: src/exporter/http_server.cc
Comment (cbodley) [reply src/exporter/http_server.cc:152]: > But yeah we planned not to make it the part of the ceph daemons so I don't think we'll need ceph.conf for this daemon

why not support ceph.conf? is there a reason to avoid fetching config from the monitor? there may be configurables (like polling intervals) that we want to change at runtime, and command line arguments can't do that

File: src/exporter/CMakeLists.txt
Comment (cbodley) [reply src/exporter/CMakeLists.txt:4]: ceph_exporter.cc is under src/ but the others are under src/exporter/

File: src/exporter/http_server.cc
Comment (epuertat) [reply src/exporter/http_server.cc]: @cbodley shouldn't that be a 404?

File: src/exporter/http_server.cc
Comment (cbodley) [reply src/exporter/http_server.cc:152]: > It could be through ceph.conf, ceph config, or whatever

this exporter is a ceph daemon, and i think it's easiest to use ceph's existing primitives for configuration, argument parsing, logging, etc

with `common/ceph_argparse.h`, any ceph.conf variables can be provided on the command line. we could choose whether or not connect to ceph-mon for config, but i think that's worth doing too for runtime changes

we'll also need logging to monitor the daemon and debug any issues, and we should do that via `CephContext` and `common/dout.h`

> but probably the less smart-guessing we do here, the better (if no interface specified, the usual behaviour on webservers is to bind to all IPv4 interfaces or 0.0.0.0).

agreed, this is consistent with the `exporter_addr` suggestion above

File: src/cephadm/cephadm
Comment (adk3798) [inline src/cephadm/cephadm:4419]: is there a better way we can split up this line than a single right paren on its own line?

File: src/pybind/mgr/cephadm/module.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:96]: Why not a container image for this daemon type instead of changing the default ceph image we're using?

File: src/pybind/mgr/cephadm/module.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:98]: why are we lowering the prometheus and node-exporter version?

File: src/pybind/mgr/cephadm/module.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:1334]: is this also meant to be a temporary thing? Why would we not be able to figure out the image for this exporter? Unless it isn't meant to be containerized?

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:1025]: these don't need any config at all? They don't need any info on where other daemons are or anything?

File: src/pybind/mgr/cephadm/services/monitoring.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/monitoring.py:311]: `mgr_scrape_list` is meant to be the list of host/port for the mgr daemons. Can we use a different variable for gathering the list of ceph-exporters?

File: src/pybind/mgr/orchestrator/_interface.py
Comment (adk3798) [inline src/pybind/mgr/orchestrator/_interface.py:740]: do we also want this daemon type to be added to the upgrade order? If it's going to be part of th ecpeh image it should definitely be added there.
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/utils.py#L24-L30

File: src/cephadm/cephadm
Comment (adk3798) [inline src/cephadm/cephadm:510]: what's this based on? Are we seeing issues with prometheus and need to up its memory limit to fix it? 8GB is more than most daemons get.

File: src/cephadm/cephadm
Comment (adk3798) [inline src/cephadm/cephadm:5774]: any reason for the hardcoded 0/0 uid/gid rather than using `extract_uid_gid` like the other daemons? Does it absolutely need to be 0/0 for this to woek?

**Review by adk3798 — COMMENTED** [review]
General structure of the cephadm stuff to deploy the container looks good. Mostly some questions about things.

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm:510]: Sorry this change needs to be reverted, will fix it!

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm:5774]: Yeah for now I think it needs (0,0) , but I think as now the plan is to make it a ceph daemon (in the next iteration) it'll be easy to fetch the uid,gid just like other other ceph daemons

File: src/pybind/mgr/orchestrator/_interface.py
Comment (avanthakkar) [reply src/pybind/mgr/orchestrator/_interface.py:740]: Yes, it'll be a part of the default ceph iage itself. Yeah that cchange I've left for later as currently we're using a image for ceph for testing purpose, but once feature branch is merged in master it'll be the part of ceph image.

File: src/pybind/mgr/cephadm/services/monitoring.py
Comment (avanthakkar) [reply src/pybind/mgr/cephadm/services/monitoring.py:311]: OK, sure!

File: src/pybind/mgr/cephadm/services/cephadmservice.py
Comment (avanthakkar) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:1025]: The configuration will be introduced in the next iteration. port, addrs, cert,key,etc.. will be the config we're planning to introduce for this daemon

File: src/pybind/mgr/cephadm/module.py
Comment (avanthakkar) [reply src/pybind/mgr/cephadm/module.py:98]: This must be a mistake!

File: src/pybind/mgr/cephadm/module.py
Comment (avanthakkar) [reply src/pybind/mgr/cephadm/module.py:96]: We're not merging this into master directly, as the image will keep on changing in this branch as long as we introduce new changes on C++ side and once feature branch is merged in master it'll be the part of default ceph image.

File: src/exporter/http_server.cc
Comment (avanthakkar) [reply src/exporter/http_server.cc:152]: Yes dns_lookup() is just  a temporary thing for now, as in the next iteration we'll introduce ceph.conf for this daemon , which will include exporter address configuration as well.

File: src/cephadm/cephadm
Comment (avanthakkar) [reply src/cephadm/cephadm:908]: Done

**Comment by avanthakkar** [comment]
jenkins test api

File: src/exporter/http_server.cc
Comment (cbodley) [reply src/exporter/http_server.cc:93]: the http thread is reading from `DaemonMetricCollector::metrics`, and the collector thread is writing to it. that's a data race, so it either needs to be fixed with atomics or locks

`std::string` isn't atomic, so both the reads and the writes to `DaemonMetricCollector::metrics` need to be protected by a mutex

**Review by pereman2 — APPROVED** [review]

**Comment by pereman2** [comment]
We'll keep track of the whole progress in https://github.com/ceph/ceph/pull/46213.

**Pull request merged**
